### PR TITLE
See BcBio Run

### PR DESCRIPTION
**Pull request by danbills**
a melange of small things:

* introduce WomLong type
* allow runtime processes to look for/ export files in "inputs" folder (not just "execution")
* Execute JSON stringify on values that are mapped to an execution environment (e.g. docker)
* Properly scale ResourceRequirement/Memory from MB to bytes

## Todo

Figure out inputbinding precedence.

### CODE REVIEW

File: cwl/src/main/scala/cwl/internal/GigabytesToBytes.scala [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+  implicit def long = at[Long] {
+    l =>
+      val value = l * toMebibytesMultiplier
+      Coproduct[cwl.ResourceRequirementType](value)
+  }
+
+  implicit def string = at[String] {
+    s =>
+      //TODO: Scale this by multiplier
```
Comment (danbills) [inline cwl/src/main/scala/cwl/internal/GigabytesToBytes.scala]: this should be elevated to a bug/issue as it's going to bite someone someday

Comment (danbills) [reply cwl/src/main/scala/cwl/internal/GigabytesToBytes.scala]: https://github.com/broadinstitute/cromwell/issues/3382

File: cwl/src/main/scala/cwl/MyriadInputTypeToSortedCommandParts.scala [hunk 2]
```diff
@@ -172,8 +172,14 @@ object MyriadInputInnerTypeToSortedCommandParts extends Poly1 {
                   .append(ias.inputBinding, Coproduct[StringOrInt](index))
 
               // Even if the item doesn't have an explicit input binding, it should appear in the command so create a default empty one
-              val itemInputBinding = ias.inputBinding.orElse(Option(InputCommandLineBinding.default))
-              // Fold over the item type fo each array element
+              //there is an explicit input binding!
+              //TODO: Figure out precedence order here!
```
Comment (mcovarr) [inline cwl/src/main/scala/cwl/MyriadInputTypeToSortedCommandParts.scala]: and for me at least also figure out what that even means

File: cwl/src/main/scala/cwl/internal/EcmaScriptEncoder.scala [hunk 3]
```diff
@@ -5,7 +5,7 @@ import EcmaScriptUtil.{ECMAScriptVariable, ESArray, ESObject, ESPrimitive}
 import cats.data.Validated.Valid
 import common.validation.ErrorOr.ErrorOr
 import cwl.{Directory, File}
-import wom.values.{WomArray, WomBoolean, WomCoproductValue, WomFile, WomFloat, WomGlobFile, WomInteger, WomMap, WomMaybeListedDirectory, WomMaybePopulatedFile, WomObjectLike, WomOptionalValue, WomSingleFile, WomString, WomUnlistedDirectory, WomValue}
+import wom.values.{WomArray, WomBoolean, WomCoproductValue, WomFile, WomFloat, WomGlobFile, WomInteger, WomLong, WomMap, WomMaybeListedDirectory, WomMaybePopulatedFile, WomObjectLike, WomOptionalValue, WomSingleFile, WomString, WomUnlistedDirectory, WomValue}
```
Comment (mcovarr) [inline cwl/src/main/scala/cwl/internal/EcmaScriptEncoder.scala]: preexisting but this is what `_` was born to do

File: wom/src/main/scala/wom/types/WomLongType.scala [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import spray.json.{JsNumber, JsString}
+import wom.values.{WomInteger, WomLong, WomString}
+
+case object WomLongType extends WomPrimitiveType {
+  val toDisplayString: String = "Long"
+
+  override protected def coercion = {
+    case i: Long => WomLong(i)
+    case i: Integer => WomLong(i.toLong)
+    case n: JsNumber if n.value.isValidLong => WomInteger(n.value.intValue())
```
Comment (mcovarr) [inline wom/src/main/scala/wom/types/WomLongType.scala]: That won't end well if it doesn't fit in an int. And shouldn't all this stuff end up as `WomLong`s?

Comment (Horneth) [reply wom/src/main/scala/wom/types/WomLongType.scala]: ToL:  some unit tests to validate those coercions would be nice too

File: cwl/src/main/scala/cwl/CwlWomExpression.scala [hunk 5]
```diff
@@ -83,12 +72,20 @@ object InitialWorkDirFileGeneratorExpression {
...
     implicit val caseExpressionDirent: Case.Aux[ExpressionDirent, InitialWorkDirFileEvaluator] = {
       at { expressionDirent =>
-        (ioFunctionSet, parameterContext, expressionLib) => {
-          val entryEvaluation = ExpressionEvaluator.eval(expressionDirent.entry, parameterContext, expressionLib)
+        (ioFunctionSet, unmappedParameterContext, mappedInputValues, expressionLib) => {
+
+          val entryEvaluation =
+            expressionDirent.entry match {
+                //we need to catch this special case to feed in "value-mapped" input values
+              case expr@Expression.ECMAScriptExpression(exprString) if exprString.value == "$(JSON.stringify(inputs))" =>
```
[the remark this answers is not in the document]
Comment (danbills) [reply cwl/src/main/scala/cwl/CwlWomExpression.scala]: yeah.  I'll put a little lipstick on this pig but it works so chalk one up for pragmatism

File: backend/src/main/scala/cromwell/backend/backend.scala [hunk 6]
```diff
@@ -73,8 +73,11 @@ case class BackendWorkflowDescriptor(id: WorkflowId,
   */
 case class BackendConfigurationDescriptor(backendConfig: Config, globalConfig: Config) {
 
-  lazy val backendRuntimeConfig = if (backendConfig.hasPath("default-runtime-attributes"))
-    Option(backendConfig.getConfig("default-runtime-attributes")) else None
+  lazy val backendRuntimeConfig =
+    if (backendConfig.hasPath("default-runtime-attributes"))
+      Option(backendConfig.getConfig("default-runtime-attributes"))
+    else
+      None
```
[the remark this answers is not in the document]
Comment (danbills) [reply backend/src/main/scala/cromwell/backend/backend.scala]: I'm going to leave it as is as it's retrieving a `Config`, not a type which is where I would expect the Ficus use case.

Of course I'd love to type-ify all our configs but I don't think this particular use case is standing out from the crowd.

File: cwl/src/main/scala/cwl/CommandLineTool.scala [hunk 7]
```diff
@@ -231,11 +231,11 @@ case class CommandLineTool private(
     }
 
     // This seems like it makes sense only for CommandLineTool and hence is not abstracted in Tool. If this assumption is wrong it could be moved up.
-    val adHocFileCreations: Set[WomExpression] = (for {
+    val adHocFileCreations: Set[MappedAndUnmappedInputs] = (for {
```
[the remark this answers is not in the document]
Comment (danbills) [reply cwl/src/main/scala/cwl/CommandLineTool.scala]: Yeah I struggled with this and ultimately couldn't think of something really good, although I like the `Container/host` idea.  I'll give it another go.

**Pull request merged**